atomic_reactor/dirs.py

```diff
@@ -0,0 +1,227 @@
+        """
+        platform: Platform = sorted(self.platforms, key=lambda item: item.value)[0]
+        return BuildDir(self.path / platform.value, platform)
+
+    def for_each(self, action: Callable[[BuildDir], Any]) -> Dict[Platform, Any]:
+        """Apply an action on every platform-specific directory.
+
+        The action callable will be applied to the platform-specific
+        directories in the order of the platforms specified when initializing
+        this RootBuildDir object. The caller should not depend on or assume the
+        order how action is applied.
+
+        The action callable might raise errors, that will cause the
+        ``for_each`` terminates immediately and the raised error is propagated
+        to the caller. As a result, the action will not be applied to the rest
+        of the platforms.
+
+        :param action: a callable object that will be applied on every
+            platform-specific directory. This callable must accept one single
+            argument in BuildDir type, and it can return data in any type.
+        :type action: Callable
+        :return: a mapping from platform to the value returned from the
+            function which is called for that platform.
+        :rtype: dict[Platform, any]
+        """
+        results = {}
+        for platform in self.platforms:
+            build_dir = BuildDir(self.path / platform.value, platform)
+            results[platform] = action(build_dir)
+        return results
+
+    def for_all_copy(self, action: Callable[[BuildDir], Iterable[Path]]) -> List[Path]:
+        """Ensure created files are present in all platform-specific directories.
+
+        :param action: a callable that creates files in a given build directory
+            and returns the files it created. The action accepts one single
+            argument in type BuildDir, and returns an iterable object that
+            yields paths of the created files.
+        :type action: callable
+        :return: the list of absolute paths of the created files in all build
+            directories.
+        :rtype: list[pathlib.Path]
+        """
+        first_platform: Platform = self.platforms[0]
+        build_dir = self.path / first_platform.value
+        created_files = action(BuildDir(build_dir, first_platform))
+
+        # Ensure the created files' path is the absolute path, and make some
+        # checks.
+        the_new_files: List[Path] = []
+        filepath: Path
+        for filepath in created_files:
+            if not filepath.is_absolute():
+                filepath = filepath.resolve()
+            if filepath.is_dir():
+                raise ValueError(f"{filepath} is a directory. A normal file is expected.")
+            if not str(filepath).startswith(str(build_dir)):
+                raise ValueError(
+                    f"File must be created inside the build directory. "
+                    f"File {filepath} is not allowed."
+                )
```

Comment:
Big +1 for this, but I think the check could be a bit more robust:
```python
build_dir in filepath.parents
# or
try:
    filepath.relative_to(build_dir)
except ValueError:
    ...
```
The current code may miss some edge cases like
```python
"/foobar/some-file".startswith("/foo")
```

Reply:
Ooh, I didn't know about [is_relative_to](https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.is_relative_to), nice!

But it's new in 3.9, so we can't use it yet :cry:

Reply:
Oh, I didn't notice that. I'll change it to `relative_to`. :sweat_smile:

Reply:
Done. Updated to use `relative_to`.